client/state/initial-state.js

```diff
@@ -194,7 +225,7 @@ export function persistOnChange( reduxStore ) {
 			const _timestamp = Date.now();
 
 			const storeTasks = map( serializedState.get(), ( data, storageKey ) =>
-				persistentStoreState( reduxStateKey, storageKey, data, _timestamp )
+				persistentStoreState( reduxStateKey, storageKey, data, _timestamp, updateCache )
```

Comment:
When I was testing this patch locally, my root state was never persisted to IndexedDB -- I always got only the `redux-state-1234:signup` key saved.

It's because this `map()` call has a bug: all the `persistentStoreState` calls are async tasks that run in parallel, and as they all read state from storage and write it back, they keep overwriting each other's changes:
1. Task A reads state from storage
2. Task B reads state from storage
3. Task A writes updated state to storage, but only after B's read
4. Task B writes updated state to storage, but overwrites A's changes

With the `calypso-state` umbrella key, the updates need to be orchestrated more carefully. Such bugs can be avoided by storing the `redux-state-*` bits separately, as we did until now.

Reply:
Excellent catch, @jsnajdr! I'll get started on modifying this PR to work on top of #38320, then.